mlx/primitives.h

```diff
@@ -980,6 +980,32 @@ class Negative : public Primitive {
   void eval(const std::vector<array>& inputs, array& out);
 };
 
+class Norm : public Primitive {
```

Comment:
This is not really a core primitive. We might want to to something as 
```
namespace mlx::linalg {
  using namespace mlx::core;
  
  class Norm : public Primitive {
  public:
    explicit Norm(
        Stream stream,
        const std::variant<std::monostate, int, std::string>& ord,
        const std::vector<size_t>& axes)
        : Primitive(stream), ord_(ord), axes_(axes){};

    void eval_cpu(const std::vector<array>& inputs, array& out) override;
    void eval_gpu(const std::vector<array>& inputs, array& out) override;

    std::pair<array, int> vmap(
        const std::vector<array>& inputs,
        const std::vector<int>& axes) override;

    DEFINE_GRADS()
    DEFINE_PRINT(NORM)
    DEFINE_DEFAULT_IS_EQUIVALENT()

  private:
    std::variant<std::monostate, int, std::string> ord_;
    std::vector<size_t> axes_;

    void eval(const std::vector<array>& inputs, array& out);
  };
}
```

Reply:
Or more importantly why does it even need to be a primitive? Why not simply do a power and reduce?

Reply:
I do not think this is a primitive.

Reply:
In fact, making it a primitive complicates the implementation since we need to implement vmap, jvp, vjp ...

Reply:
Exactly. The comment was meant towards @j-csc mostly to point him towards implementing it with `x.power(p).sum(axis=ax).power(1/p)` (of course the equivalent in c++ which is only slightly more verbose).

Reply:
Happy to make the change, or even better just take it off and let @gboduljak push up his implementation! 

The thought process was more on attempting to standardize all linalg module function impls, and use lapack/accelerate as much as possible (which I’m not even sure if primitives are a must for this)

Reply:
Was hesitant on the standardization that’s why I wanted to push it up and get some feedback 😅

Reply:
I will implement the version of norm as a method, as @angeloskath suggested.

Reply:
They are but we want to avoid Primitive explosion. Maybe QR requires a primitive to run fast but norm definitely not. Would we eke out slightly more performance with a kernel? Maybe but it will be insignificant while adding a large maintenance overhead.

Reply:
No that's great! We appreciate the love and the PRs :-) .

Reply:
Yea was having same thoughts tbh these primitives are hard to maintain but was unsure. Will look at other functions when I get back but thanks guys this was really constructive.

also in the future would you prefer draft prs for questions like these or in issues

Reply:
No preference really. I think this was nice as a draft PR because you can easily show what you have in mind in a PR. In an issue you 'd have to do kinda the same and then copy paste it into code snippets.

Reply:
> Maybe QR requires a primitive to run fast
this definitively does. i have it as a primitive my in my PR and it calls LAPACK routines.

Reply:
the primitive ADT mainly has eval_{cpu|gpu} signature that returns a single output. most factorization methods return more than one output. a pair perhaps? or a vector of array. 

what's the preferable way to do this without breaking the current convention? cc @awni